src/sc/sidechain.cpp

```diff
@@ -782,107 +781,98 @@ void CSidechain::UpdateScFees(const CScCertificateView& certView, int blockHeigh
         {
             const auto& entry = scFees.front();
             LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
-                __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
+                __func__, __LINE__, entry->forwardTxScFee, entry->mbtrTxScFee);
             scFees.pop_front();
         }
     }
     // v2 sidechains
     else {
-        auto scFeeIt = std::find_if(scFees_v2.begin(), scFees_v2.end(),
-            [&blockHeight](Sidechain::ScFeeData_v2 scFeeElem) { return scFeeElem.submissionHeight == blockHeight; });
+        auto scFeeIt = std::find_if(scFees.begin(), scFees.end(),
+            [&blockHeight](const std::shared_ptr<Sidechain::ScFeeData> & scFeeElem) {
```

Comment:
What about adding a check on the cast result inside the lambda?
Something like:

```
[&blockHeight](const std::shared_ptr<Sidechain::ScFeeData> & scFeeElem) {
    std::shared_ptr<Sidechain::ScFeeData_v2> casted_scFeeElem = std::dynamic_pointer_cast<Sidechain::ScFeeData_v2>(scFeeElem);
    if (casted_entry)
        return casted_entry.get()->submissionHeight == blockHeight;
    return false;
}
```

I see there is a check on the non ceasability of the SC, but maybe this option can be more robust in case of future extensions/developments

Reply:
As you also notice, if we are executing this, then it means it's a non ceasing sidechain, which means we *must* have `ScFeeData_v2`. If something goes wrong/changes in the future, I'd rather not being silent about unexpected conditions, but instead scream out loud immediately.

On the other hand, I agree we should do better than undefined behavior. I propose an `assert(casted_entry)`, is that ok with you guys?

Reply:
Yes, I think it's a good solution.
Returning `false` may not be enough if dynamic casting fails.